lambda/headless/lambda_handler.js

```diff
@@ -0,0 +1,177 @@
+      // useful to know the start/end from Lambda's vantage point.
+      var end_time = new Date().getTime() / 1000;
+      var duration = end_time - start_time;
+
+      var response = {
+        lambda: {
+          log_group_name: context.logGroupName,
+          log_stream_name: context.logStreamName,
+          request_id: context.awsRequestId,
+          memory_limit: context.memoryLimitInMB,
+          start_time: start_time,
+          end_time: end_time,
+          measured_duration: duration
+        }
+      }
+
+      if (err) {
+        if (err instanceof Error)
+          response.error = err.stack;
+        else
+          response.error = err;
+      }
+
+      if (data === undefined) {
+        response.data = null;
+        if (!err)
+          response.error = "Data came back undefined for some reason.";
+      }
+      else
+        response.data = data;
+
+      // TODO: JSON datetime sanitization, like the Python handler does.
+      callback(null, response);
+    }
+  );
+};
+
+
+const puppeteer = require('puppeteer');
+var path = require('path');
+var fs = require('fs');
+var tar = require('tar');
+
+const setupLocalChrome = () => {
+  return new Promise((resolve, reject) => {
+    fs.createReadStream(localChromePath)
+    .on('error', (err) => reject(err))
+    .pipe(tar.x({
+      C: setupChromePath,
+    }))
+    .on('error', (err) => reject(err))
+    .on('end', () => resolve());
+  });
+};
+
+const localChromePath = path.join('headless_shell.tar.gz');
+const setupChromePath = path.join(path.sep, 'tmp');
+const executablePath = path.join(
+    setupChromePath,
+    'headless_shell'
+);
+
+
+const chromeOptions = [
+  // Resolves error:
+  //    error when launch(); No usable sandbox! Update your kernel
+  '--no-sandbox',
+
+  // Resolves error:
+  //    error when launch(); Failed to load libosmesa.so
+  '--disable-gpu',
+
+  // Resolves error: freeze when newPage()
+  '--single-process'
+];
+
+
+// Async function to load Chrome from the Lambda container.
+var getBrowser = (() => {
+  let browser;
+  return async () => {
+    // idempotence
+    if (typeof browser === 'undefined' || !await isBrowserAvailable(browser)) {
+
+      // extract .tar.gz to tmp/
+      await setupLocalChrome();
+
+      // load the browser from its extracted path
+      browser = await puppeteer.launch({
+        headless: true,
+        executablePath: executablePath,
+        args: chromeOptions,
+        ignoreHTTPSErrors: true,
+
+        // puppeteer 1.1.0 broke dumpio. Should be fixed soon.
+        // dumpio: true
+      });
+
+      var version = await browser.version()
+      console.log("Launched: " + version);
+    }
+    return browser;
+  };
+})();
+
+const isBrowserAvailable = async (browser) => {
+  try {
+    await browser.version();
+  } catch (e) {
+    console.log(e); // not opened etc.
+    return false;
+  }
+  return true;
+};
+
+
+/** local test run **/
+
+if (process.env.TEST_LOCAL) {
```

Comment:
This whole section below here is just to allow quicker (local) execution of the Lambda handler code-path without needing to upload the whole ~50MB function to Lambda and test it up there. It specifies a tiny CLI API to handler.js that is not ever used in actual Lambda-based execution of functions.